### PR TITLE
fix(issue #2205): FK is already defined as the defined field

### LINKED ISSUE #2205: code generation error when using bidirectional edges with StorageKey
- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/ent/ent/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

functions at generated codes are redeclared whern using bidirectional edges with StorageKey.

## Expected Behavior 🤔

functions should be generated w/o problem, or just return error message when generating code via `go generate ./...`

## Steps to Reproduce 🕹

Steps:

1. Write schema codes below.

```

// ConditionGroup holds the schema definition for the ConditionGroup entity.
type ConditionGroup struct {
	ent.Schema
}

// Annotations of the ConditionGroup.
func (ConditionGroup) Annotations() []schema.Annotation {
	return []schema.Annotation{
		entsql.Annotation{Table: "condition_group"},
	}
}

// Fields of the ConditionGroup.
func (ConditionGroup) Fields() []ent.Field {
	return []ent.Field{
		field.Int64("id").
			Unique().
			Immutable(),
	}
}

// Edges of the ConditionGroup.
func (ConditionGroup) Edges() []ent.Edge {
	return []ent.Edge{
		edge.To("condition", Condition.Type).
			StorageKey(edge.Column("condition_group_id")),
	}
}


// Condition holds the schema definition for the Condition entity.
type Condition struct {
	ent.Schema
}

// Annotations of the Condition.
func (Condition) Annotations() []schema.Annotation {
	return []schema.Annotation{
		entsql.Annotation{Table: "condition"},
	}
}

// Fields of the Condition.
func (Condition) Fields() []ent.Field {
	return []ent.Field{
		field.Int64("id").
			Unique().
			Immutable(),
		field.Int64("condition_group_id"),
		field.String("value"),
	}
}

// Edges of the Condition.
func (Condition) Edges() []ent.Edge {
	return []ent.Edge{
		edge.From("condition_group", ConditionGroup.Type).
			Ref("condition").
                         // Field("condition_group_id").
			Unique(),
	}
}
```

2. Generate stub codes using `go generate ./...`

3. `SetConditionGroupID` function in stub codes are redeclared, so I can't run any generated code.

4. when try to run test code, error messages are shown.

```
lambert@lambertui-MacBookPro storagekey % go test -v .
# entgo.io/ent/entc/integration/storagekey/ent
ent/condition_create.go:42:28: (*ConditionCreate).SetConditionGroupID redeclared in this block
        /Users/lambert/oss/ent/entc/integration/storagekey/ent/condition_create.go:24:6: previous declaration
ent/condition_update.go:51:28: (*ConditionUpdate).SetConditionGroupID redeclared in this block
        /Users/lambert/oss/ent/entc/integration/storagekey/ent/condition_update.go:32:6: previous declaration
ent/condition_update.go:247:32: (*ConditionUpdateOne).SetConditionGroupID redeclared in this block
        /Users/lambert/oss/ent/entc/integration/storagekey/ent/condition_update.go:228:6: previous declaration
ent/mutation.go:245:29: (*ConditionMutation).SetConditionGroupID redeclared in this block
        /Users/lambert/oss/ent/entc/integration/storagekey/ent/mutation.go:153:6: previous declaration
ent/mutation.go:260:29: (*ConditionMutation).ConditionGroupID redeclared in this block
        /Users/lambert/oss/ent/entc/integration/storagekey/ent/mutation.go:159:6: previous declaration
FAIL    entgo.io/ent/entc/integration/storagekey [build failed]
```

## Your Environment 🌎

| Tech        | Version |
| ----------- | ------- |
| Go          | 1.17.2  |
| Ent         | 0.9.1   |
| Database    | MySQL   |
| Driver      | https://github.com/go-sql-driver/mysql        |

### PR DESCRIPTION
**Pull request by TcMits**
Hello @a8m

Fixed #2205

### CODE REVIEW

**Review by a8m — COMMENTED** [review]
Hey @TcMits, and thanks for the contribution. 

I want to review this but there are too many unrelated changes in this PR. Please, clean them and tag me when it's ready. Thanks 🙏

**Comment by TcMits** [comment]
Hey @a8m , i removed unrelated changes. Sorry i forgot turning off auto formatter. 😢

**Comment by TcMits** [comment]
Hey @a8m, i removed unrelated changes. it's ready to review. Thanks 😄

File: entc/gen/type.go [hunk 1]
```diff
@@ -708,16 +708,33 @@ func (t *Type) setupFKs() error {
 				}
 			}
 		}
-		// Special case for checking if the FK is already defined as the ID field (see issue 1288).
-		if key, _ := e.StorageKey(); key != nil && len(key.Columns) == 1 && key.Columns[0] == refid.StorageKey() {
-			fk.Field = refid
-			fk.UserDefined = true
+		// Special case for checking if the FK is already defined as the defined field (see issue 1288, 2205).
```
Comment (a8m) [inline entc/gen/type.go:711]: Please, add a test case to help understand what this change does. Thanks

Comment (TcMits) [reply entc/gen/type.go:711]: Hi @a8m, i added a test case for this. Hope it helps 😄

Comment (TcMits) [reply entc/gen/type.go:711]: Hey @a8m, i added a test case for this. Hope it helps 😄

Comment (a8m) [reply entc/gen/type.go:711]: Thanks, but integration tests are also necessary (see `entc/integration`).

Comment (TcMits) [reply entc/gen/type.go:711]: added @a8m. 😄

File: entc/integration/edgefield/ent/schema/user.go [hunk 2]
```diff
@@ -36,6 +36,7 @@ func (User) Fields() []ent.Field {
 func (User) Edges() []ent.Edge {
 	return []ent.Edge{
 		edge.To("pets", Pet.Type),
+		edge.To("previous_pets", Pet.Type).StorageKey(edge.Column("previous_owner_id")),
```
Comment (a8m) [inline entc/integration/edgefield/ent/schema/user.go:39]: Why is StorageKey defined here? It should be defined on the schema that holds the foreign key using `Field` instead. Ent should throw an error in this case. 

Do you have any other case that reproduces it?

**Comment by a8m** [comment]
Hey! Please, see my comment on [#2205](https://github.com/ent/ent/issues/2205#issuecomment-1295976877) and let me know if there's anything else that this PR solves or if you think this behavior is necessary (for detecting the reference in the "other" side).

**Comment by TcMits** [comment]
Hey @a8m, 

You're right, throwing error in this case ( #2205 ) is the fastest way to solve problem.  But following the [docs](https://entgo.io/docs/schema-edges#storagekey):
> By default, Ent configures edge storage-keys by the edge-owner (the schema that holds the edge.To), and not the by back-reference (edge.From). This is because back-references are optional and can be removed.

So i think in most cases, we should define `storage-keys` in `edge-owner`. Then we can remove back-reference without any problems ( affect relations ). Sorry if i missed somethings. 😄

I tried default behavior of Ent when i removed back-references ( in edgefield integration test ), output struct of `pet.go` file:
```go
type Pet struct {
        ...
	PreviousOwnerID int `json:"previous_owner_id,omitempty"`
        ...
	previous_owner_id *int // didn't expect this
}
```
I added a [commit](https://github.com/ent/ent/pull/3032/commits/68035817585965610f0f8bd10e4119aebf436b2a) to change default behavior of Ent. Please tell me if i'm wrong.

Thanks.

**Comment by a8m** [comment]
Understand your point @TcMits, and it's appreciated. But if the edge field is defined in the referenced schema, I expect the users to use it / declare it there. I find it cleaner, more explicit, and works well with other integrations, such as EntGQL.

**Comment by a8m** [comment]
Again, your contribution to Ent is highly appreciated. I see it a bit differently when it comes to the schema definition.

**Comment by TcMits** [comment]
> Again, your contribution to Ent is highly appreciated. I see it a bit differently when it comes to the schema definition.

Thanks. @a8m 

> Understand your point @TcMits, and it's appreciated. But if the edge field is defined in the referenced schema, I expect the users to use it / declare it there. I find it cleaner, more explicit, and works well with other integrations, such as EntGQL.

Okay, i will close this PR.

**Comment by a8m** [comment]
❤️

**Pull request closed without merging**